src/java/fr/paris/lutece/plugins/limitconnectedusers/service/LimitSessionService.java

```diff
@@ -88,4 +89,27 @@ public void setNbMaximumUsersReached( boolean _bNbMaximumUsersReached )
     {
         this._bNbMaximumUsersReached = _bNbMaximumUsersReached;
     }
+    
+        
+    /**
+     * get absolute server Session number
+     * @return session nb
+     */   
+    public int getNbSessions() {
+        return _nbSessions;
+    }
+
+    /**
+     * add a session
+     */
+    public void addSession( ) {
+        this._nbSessions ++ ;

```

Comment:
This does not seem thread safe. What happens if two sessions are created at the same time ?
Maybe _nbSessions should be of type java.util.concurrent.atomic.AtomicInteger ?

Reply:
This problem should only occur at fist call, no ?
We could synchronized the LimitSessionService getService(  ) method, but it could take too much execution time.
Maybe we should use the Initialization-on-demand holder  pattern ?
https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom

Reply:
This can happen on any call to addSession( ) or removeSession( ) if tow such call happen concurrently. See https://docs.oracle.com/javase/tutorial/essential/concurrency/interfere.html for a discussion of that. Synchronizing access to the getService() method does not do anything for that. It does not prevent concurrent calls to its methods.

I see also that this class does not seem to protect _listSessionsActive from concurrent access either. Maybe that should be another JIRA...

Reply:
The rest of the code was already not thread safe :( it calls HashSet::add and HashSet::contains which are not thread safe

![](http://bholley.net/images/posts/thistall.jpg)

Reply:
well, the whole plugin is a DOS facilitator, so...
But I think at least new code should be correct.

Reply:
This problem should only occur at fist call, no ?
    We could synchronized the LimitSessionService getService( ) method, but it could take too much execution time.
    Maybe we should use the Initialization-on-demand holder pattern ?
    https://en.wikipedia.org/wiki/Initialization-on-demand_holder_idiom

No rémi is not talking about initialization.